monitors/ContractMonitor.js

```diff
@@ -0,0 +1,173 @@
+    this.syntheticCurrencySymbol = "UMATEST";
+
+    // TODO: get the decimals of the collateral currency and use this to scale the output appropriately for non 1e18 colat
+    this.formatDecimalString = createFormatFunction(this.web3, 2);
+  }
+
+  // Calculate the collateralization Ratio from the collateral, token amount and token price
+  // This is cr = [collateral / (tokensOutstanding * price)] * 100
+  calculatePositionCRPercent = (collateral, tokensOutstanding, priceFunction) => {
+    return this.web3.utils
+      .toBN(collateral)
+      .mul(this.web3.utils.toBN(this.web3.utils.toWei("1")))
+      .mul(this.web3.utils.toBN(this.web3.utils.toWei("1")))
+      .div(this.web3.utils.toBN(tokensOutstanding).mul(this.web3.utils.toBN(priceFunction.toString())))
+      .muln(100);
+  };
+
+  // Queries disputable liquidations and disputes any that were incorrectly liquidated.
+  checkForNewLiquidations = async priceFunction => {
+    Logger.debug({
+      at: "ContractMonitor",
+      message: "Checking for new liquidation events",
+      price: priceFunction.toString(),
+      lastLiquidationBlockNumber: this.lastLiquidationBlockNumber
+    });
+
+    // Get the latest liquidation information.
+    let latestLiquidationEvents = this.empEventClient.getAllLiquidationEvents();
+
+    // Get liquidation events that are newer than the last block number we've seen
+    let newLiquidationEvents = latestLiquidationEvents.filter(event => event.blockNumber > this.lastDisputeBlockNumber);
+
+    for (let event of newLiquidationEvents) {
+      // Sample message:
+      // Liquidation alert: [ethereum address if third party, or “UMA” if it’s our bot]
+      // initiated liquidation for for [x][collateral currency]of sponsor collateral
+      // backing[n] tokens - sponsor collateralization was[y] %.  [etherscan link to txn]
+      const mrkdwn =
+        createEtherscanLinkMarkdown(this.web3, networkUtils, event.liquidator) +
+        (this.umaLiquidatorAddress == event.liquidator ? " (UMA liquidator bot)" : "") +
+        " initiated liquidation for " +
+        this.formatDecimalString(event.liquidatedCollateral) +
+        " " +
+        this.collateralCurrencySymbol +
+        " of sponsor " +
+        createEtherscanLinkMarkdown(this.web3, networkUtils, event.sponsor) +
+        " collateral backing " +
+        this.formatDecimalString(event.tokensOutstanding) +
+        " " +
+        this.syntheticCurrencySymbol +
+        " tokens. Sponsor collateralization was " +
+        this.formatDecimalString(
+          this.calculatePositionCRPercent(event.liquidatedCollateral, event.tokensOutstanding, priceFunction)
+        ) +
+        "%. tx: " +
+        createEtherscanLinkMarkdown(this.web3, networkUtils, event.transactionHash);
+
+      Logger.info({
+        at: "ContractMonitor",
+        message: "Liquidation Alert 🧙‍♂️!",
+        mrkdwn: mrkdwn
+      });
+    }
+    this.lastLiquidationBlockNumber = await this.web3.eth.getBlockNumber();
```

Comment:
I think if you compute the block number here, you may miss events because the block number may have changed between when the client made the event query and now. Can you pull out the client's lastBlockNumber and use that? Another option would just be to take the highest block number from any event you've gotten from the client.

It might be nice to add a test for this case too. In the test, you'd want the client to update, then fire off new liquidations, then call this function. IIUC, right now, the liquidations that fall in between would be skipped and get no alerts.

Reply:
ye, good catch. I had actually realized this mistake and corrected it in the next PR on wallet monitoring but had forgotten to circle back and fix it here. Thanks for this!

I will add a test once the winston spy PR is in next 🍡